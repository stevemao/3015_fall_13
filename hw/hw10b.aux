\relax 
\@writefile{toc}{\contentsline {section}{\numberline {1}Modeling Finite Functions as Lists}{1}}
\@writefile{toc}{\contentsline {subsection}{\numberline {1.1}{\tt  {update :: Eq a => (a,b) -> FinFun a b -> FinFun a b}}}{1}}
\@writefile{toc}{\contentsline {subsection}{\numberline {1.2}\tt  {{functional :: Eq a => [(a,b)] -> Bool}}}{1}}
\@writefile{toc}{\contentsline {subsection}{\numberline {1.3}\tt  {{domain :: (Eq a, Eq b) => FinFun a b -> [a]}}}{1}}
\@writefile{toc}{\contentsline {subsection}{\numberline {1.4}\tt  {range :: (Eq a, Eq b) => FinFun a b -> [b]}}{2}}
\@writefile{toc}{\contentsline {subsection}{\numberline {1.5}\tt  {{apply :: (Eq a, Eq b) => FinFun a b -> a -> Maybe b}}}{2}}
