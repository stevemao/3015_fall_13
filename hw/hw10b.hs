 -- Finite Functions from a to b are modelled  by the type Set (a,b) 

data FinFun a b = FF [(a,b)]
 
 -- We display finite functions as {a->b,c->d,e->f} by instantiating the show function as follows:

instance (Show a, Show b) => Show (FinFun a b) where
    show (FF m) = wrap "{" "}" (comma_sep (map (\s -> concat (map (\c -> if c == ',' then "->" else [c]) s)) (map show m)))
      where wrap c1 c2 m = c1 ++ m ++ c2
            comma_sep  = foldr1 ((++) . (++ ",")) 




 -- if you only use update to build  finite functions, they will be functional

update (x,y) (FF m) = FF [] -- replace "FF []" with your code.

 
apply (FF m) x = Nothing -- replace "Nothing" with your code.

domain (FF  m) = []  -- replace "[]" with your code
range (FF m) = [] -- replace "[]" with your code

-- test a list of pairs to see if it is functional (i.e. if each domain element appears at most once.)
functional  m =  False -- replace "False" with you code.


invert (FF m) = FF (map swap  m)
  where swap (x,y) = (y,x)

c2i = FF (zip ['a'..'z'] (map fromEnum ['a'..'z']))
i2c = invert c2i
number = FF [(1,"one"), (2, "two")]
