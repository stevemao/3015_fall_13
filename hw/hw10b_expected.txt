*Main> :l hw10
:l hw10
[1 of 1] Compiling Main             ( hw10.hs, interpreted )
Ok, modules loaded: Main.
*Main> number
number
{(1->"one"),(2->"two")}
*Main> let FF m = number in m
let FF m = number in m
[(1,"one"),(2,"two")]
*Main> update (3,"three") number
update (3,"three") number
{(3->"three"),(1->"one"),(2->"two")}
*Main> update (2,"2") number
update (2,"2") number
{(2->"2"),(1->"one")}
*Main> update (1,"1") number
update (1,"1") number
{(1->"1"),(2->"two")}
*Main> c2i
c2i
{('a'->97),('b'->98),('c'->99),('d'->100),('e'->101),('f'->102),('g'->103),('h'->104),('i'->105),('j'->106),('k'->107),('l'->108),('m'->109),('n'->110),('o'->111),('p'->112),('q'->113),('r'->114),('s'->115),('t'->116),('u'->117),('v'->118),('w'->119),('x'->120),('y'->121),('z'->122)}
*Main> apply c2i 'a'
apply c2i 'a'
Just 97
*Main> apply c2i 'z'
apply c2i 'z'
Just 122
*Main> apply c2i 'A'
apply c2i 'A'
Nothing
*Main> domain number
domain number
[1,2]
*Main> range number
range number
["one","two"]
*Main> domain c2i
domain c2i
"abcdefghijklmnopqrstuvwxyz"
*Main> range c2i
range c2i
[97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122]
*Main> domain i2c
domain i2c
[97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122]
*Main> range i2c
range i2c
"abcdefghijklmnopqrstuvwxyz"
*Main> functional number
functional number

<interactive>:1:11:
    Couldn't match expected type `[(a, b)]'
           against inferred type `FinFun Integer [Char]'
    In the first argument of `functional', namely `number'
    In the expression: functional number
    In the definition of `it': it = functional number
*Main> let FF m = number in functional m
let FF m = number in functional m
True
*Main> let FF m = c2i in functional m
let FF m = c2i in functional m
True
*Main> let FF m = i2c in functional m
let FF m = i2c in functional m
True
*Main> functional [(1,0),(2,0),(3,0)]
functional [(1,0),(2,0),(3,0)]
True
*Main> map (\(x,y) -> (y,x)) [(1,0),(2,0),(3,0)]
map (\(x,y) -> (y,x)) [(1,0),(2,0),(3,0)]
[(0,1),(0,2),(0,3)]
*Main> functional it
functional it
False
*Main> map (\(x,y) -> (y,x)) [(1,0),(2,0),(3,0)]
map (\(x,y) -> (y,x)) [(1,0),(2,0),(3,0)]
[(0,1),(0,2),(0,3)]
*Main> it
it
[(0,1),(0,2),(0,3)]
*Main> it
it
[(0,1),(0,2),(0,3)]
*Main> functional it
functional it
False
*Main> it
it
False
*Main> update ('b',1) c2i
update ('b',1) c2i
{('b'->1),('a'->97),('c'->99),('d'->100),('e'->101),('f'->102),('g'->103),('h'->104),('i'->105),('j'->106),('k'->107),('l'->108),('m'->109),('n'->110),('o'->111),('p'->112),('q'->113),('r'->114),('s'->115),('t'->116),('u'->117),('v'->118),('w'->119),('x'->120),('y'->121),('z'->122)}
*Main> update ('a',1) c2i
update ('a',1) c2i
{('a'->1),('b'->98),('c'->99),('d'->100),('e'->101),('f'->102),('g'->103),('h'->104),('i'->105),('j'->106),('k'->107),('l'->108),('m'->109),('n'->110),('o'->111),('p'->112),('q'->113),('r'->114),('s'->115),('t'->116),('u'->117),('v'->118),('w'->119),('x'->120),('y'->121),('z'->122)}
*Main> update ('z',1) c2i
update ('z',1) c2i
{('z'->1),('a'->97),('b'->98),('c'->99),('d'->100),('e'->101),('f'->102),('g'->103),('h'->104),('i'->105),('j'->106),('k'->107),('l'->108),('m'->109),('n'->110),('o'->111),('p'->112),('q'->113),('r'->114),('s'->115),('t'->116),('u'->117),('v'->118),('w'->119),('x'->120),('y'->121)}
*Main> 